File: week-01/day-04/algo-anagrams-i/character_match.py
def is_character_match(string1, string2):

    #Holds character counts for each string
    character_map_1 = {}
    character_map_2 = {}

    #Remove spaces from each string
    string1 = "".join(string1.split(" "))
    string2 = "".join(string2.split(" "))

    #Populate character_map_1 with string1 characters
    for ch in string1:
        ch = ch.lower()
        if ch not in character_map_1:
            character_map_1[ch] = 1
        else:
            character_map_1[ch] += 1

    #Populate character_map_2 with string2 characters
    for ch in string2:
        ch = ch.lower()
        if ch not in character_map_2:
            character_map_2[ch] = 1
        else:
            character_map_2[ch] += 1

    #Not an anagram if a character in character_map_1 is
    #not in character_map_2 or doesn't have the same count
    for key in character_map_1:
        if (key not in character_map_2 or 
            character_map_1[key] != character_map_2[key]):
            return False

    #Not an anagram if a character in character_map_2 is
    #not in character_map_1 or doesn't have the same count
    for key in character_map_2:
        if (key not in character_map_1 or 
            character_map_1[key] != character_map_2[key]):
            return False
    
    return True

File: week-01/day-04/algo-anagrams-i/characterMatch.js
exports.isCharacterMatch = function(string1, string2) {

    //Holds character counts for each string
    let characterMap1 = {};
    let characterMap2 = {};

    //Remove spaces from each string
    string1 = string1.split(" ").join("");
    string2 = string2.split(" ").join("");

    //Populate characterMap1 with string1 characters
    for (let i = 0; i < string1.length; i++) {
        let ch = string1[i].toLowerCase();
        if (!characterMap1[ch])
            characterMap1[ch] = 1;
        else
            characterMap1[ch]++;
    }

    //Populate characterMap2 with string2 characters
    for (let i = 0; i < string2.length; i++) {
        let ch = string2[i].toLowerCase();
        if (!characterMap2[ch])
            characterMap2[ch] = 1;
        else
            characterMap2[ch]++;
    }

    //Not an anagram if a character in characterMap1 is
    //not in characterMap2 or doesn't have the same count
    for (key in characterMap1) {
        if(characterMap1[key] !== characterMap2[key])
            return false;
    }

    //Not an anagram if a character in characterMap2 is
    //not in characterMap1 or doesn't have the same count
    for (key in characterMap2) {
        if(characterMap1[key] !== characterMap2[key])
            return false;
    }
    
    return true;
};